### PR TITLE
Use kubernetes proxy for non-LoadBalancer services

### PR DESCRIPTION
**Pull request by ahmelsayed**

### CODE REVIEW

File: src/Azure.Functions.Cli/Kubernetes/KubectlHelper.cs [hunk 1]
```diff
@@ -65,7 +67,78 @@ void error(string line)
...
+
+        public static Task<Process> RunKubectlProxy(IKubernetesResource resource, int targetPort, int localPort, TimeSpan? timeout = null)
+        {
+            var kubectl = new Executable("kubectl", $"port-forward {getResourceFullName(resource)} {localPort}:{targetPort}");
+            var sbError = new StringBuilder();
+            var sbOutput = new StringBuilder();
+            var tcs = new TaskCompletionSource<Process>();
+            var deadline = DateTime.Now.Add(timeout ?? TimeSpan.FromSeconds(20));
+
+            var exitCodeTask = Task.Run(() => kubectl.RunAsync(l => output(l, false), e => output(e, true)));
```
Comment (fabiocav) [inline src/Azure.Functions.Cli/Kubernetes/KubectlHelper.cs:80]: Why the task wrapping here?

File: src/Azure.Functions.Cli/Kubernetes/KubectlHelper.cs [hunk 2]
```diff
@@ -65,7 +67,78 @@ void error(string line)
...
+            Task.Run(() => timeoutFunc());
+            return tcs.Task;
+
+            void output(string line, bool isError)
```
Comment (fabiocav) [inline src/Azure.Functions.Cli/Kubernetes/KubectlHelper.cs:84]: nit: -> `Output`

File: src/Azure.Functions.Cli/Kubernetes/KubectlHelper.cs [hunk 3]
```diff
@@ -65,7 +67,78 @@ void error(string line)
...
+                    await Task.Delay(TimeSpan.FromSeconds(1));
+                }
 
+                if (!tcs.Task.IsCompleted)
+                {
+                    tcs.TrySetException(new TimeoutException("Timedout trying to establish proxy to kubernetes api-server"));
+                }
+            }
+
+            string getResourceFullName(IKubernetesResource r)
```
Comment (fabiocav) [inline src/Azure.Functions.Cli/Kubernetes/KubectlHelper.cs:123]: nit: method casing

File: src/Azure.Functions.Cli/Kubernetes/KubectlHelper.cs [hunk 4]
```diff
@@ -65,7 +67,78 @@ void error(string line)
...
+            {
+                if (r is DeploymentV1Apps deployment)
```
Comment (fabiocav) [inline src/Azure.Functions.Cli/Kubernetes/KubectlHelper.cs:125]: Would this be simpler with a switch?

**Review by fabiocav — COMMENTED** [review]
Quick review. Just a few comments.

Any additional tests required for this? If we're trying to get this in as a fix, can we create an issue for the tests?

**Comment by ahmelsayed** [comment]
Thanks @fabiocav. I opened https://github.com/Azure/azure-functions-core-tools/issues/2420

I can probably add some unit tests for some stuff, but since most of it is about interacting with Kubernetes, I think an E2E test would give better validation. I need to setup some environment to do that though.

**Review by fabiocav — APPROVED** [review]

**Pull request merged**